opengrok-indexer/src/main/java/org/opengrok/indexer/history/GitRepository.java

```diff
@@ -568,68 +566,74 @@ boolean hasFileBasedTags() {
-        argv.add(RepoCommand);
-        argv.add("log");
-        argv.add("--format=commit:%H%nDate:%at");
-        argv.add("-n");
-        argv.add("1");
-        argv.add(tag);
-        argv.add("--");
-        
-        Executor executor = new Executor(argv, directory, interactive ?
-                RuntimeEnvironment.getInstance().getInteractiveCommandTimeout() :
-                RuntimeEnvironment.getInstance().getCommandTimeout());
-        GitTagParser parser = new GitTagParser(tag);
-        executor.exec(true, parser);
-        return parser.getEntries().first();
-    }
-
     @Override
     protected void buildTagList(File directory, boolean interactive) {
         this.tagList = new TreeSet<>();
+
+        /*
+         * Bulk log-tags command courtesy of GitHub's louie0817.
+         */
         ArrayList<String> argv = new ArrayList<>();
-        LinkedList<String> tagsList = new LinkedList<>();
         ensureCommand(CMD_PROPERTY_KEY, CMD_FALLBACK);
         argv.add(RepoCommand);
-        argv.add("tag");
-        
+        argv.add("log");
+        argv.add("--tags");
+        argv.add("--simplify-by-decoration");
+        argv.add("--pretty=%H:%at:%D:");
+
         Executor executor = new Executor(argv, directory, interactive ?
                 RuntimeEnvironment.getInstance().getInteractiveCommandTimeout() :
                 RuntimeEnvironment.getInstance().getCommandTimeout());
-        int status = executor.exec();
-
-        try {
-            // First we have to obtain list of all tags, and put it aside
-            // Otherwise we can't use git to get date & hash for each tag
-            try (BufferedReader in = new BufferedReader(new InputStreamReader(
-                    executor.getOutputStream()))) {
-                String line;
-                while ((line = in.readLine()) != null) {
-                    tagsList.add(line);
-                }
-            }
-        } catch (IOException e) {
-            LOGGER.log(Level.WARNING,
-                    "Failed to read tag list: {0}", e.getMessage());
-            this.tagList = null;
+        int status = executor.exec(true, new GitTagParser(this.tagList));
+        if (LOGGER.isLoggable(Level.FINEST)) {
+            LOGGER.log(Level.FINEST, "Read tags count={0} for {1}",
+                    new Object[] {tagList.size(), directory});
         }
 
         if (status != 0) {
             LOGGER.log(Level.WARNING,
                 "Failed to get tags for: \"{0}\" Exit code: {1}",
                     new Object[]{directory.getAbsolutePath(), String.valueOf(status)});
-            this.tagList = null;
-        }
-        
-        // Now get hash & date for each tag.
-        for (String tag : tagsList) {
-            TagEntry tagEntry = buildTagEntry(directory, tag, interactive);
-            // Reverse the order of the list
-            this.tagList.add(tagEntry);
+            // In case of partial success, do not null-out tagList here.
```

Comment:
This will basically lead to incomplete history information that cannot be restored later on. Better than the previous approach however makes me wondering if tearing the whole thing down would be in order.

Reply:
Throwing an exception would not (in current implementation) avoid the subsequent indexing stage of the project, so the end result would still be missing history information.

Reply:
True, that's a long present problem. Just wanted to mention for completeness - I am not remotely sure that terminating the whole history due to tag problem is the right thing to do.